sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,379 @@
+# Dynamic linking of device code
+
+This document describes purpose and design of dynamic linking of device code
+feature.
+
+## Background
+Sometimes users want to link device code dynamically at run time. One possible
+use case for such linkage - providing device functions via shared libraries.
+Simple source example:
+```
+// App:
+
+CGH.parallel_for<app_kernel>(/* ... */ {
+  library_function();
+});
+
+
+// Shared library:
+SYCL_EXTERNAL void library_function() {
+  // do something
+}
+```
+It is possible to manually create `sycl::program` in both app and shared
+library, then use `link` SYCL API to get a single program and launch kernels
+using it. But it is not user-friendly and it is very different from regular
+C/C++ workflow.
+
+Another possible scenario - use functions defined in pre-compiled device image
+provided by user. Example:
+```
+// a.cpp
+SYCL_EXTERNAL void foo();
+...
+parallel_for([]() { foo(); });
+
+// b.cpp
+/*no SYCL_EXTERNAL*/ void foo() { ... }
+```
+We have a `SYCL_EXTERNAL` function `foo` called from a kernel, but the
+application defined only host version of this function. Then user adds device
+image with definition of `foo` to the fat object via special option.
```

Comment:
> I would expect this application to result in an error at (static) link time.

There will be no such errors. Such errors now are emitted only by host linker and the function `foo` definition is missing only in device code (due to missing `SYCL_EXTERNAL` macro), however it is present in host code. We use llvm-link as linker of device LLVM IR, it doesn't emit errors if some function is not defined in final linked module. I also don't think we need such error emission (if there is such functionality in llvm-link) because we use undefined functions to represent SPIR-V operations in LLVM IR (please see https://github.com/intel/llvm/blob/sycl/sycl/doc/CompilerAndRuntimeDesign.md#integration-with-spir-v-format for details).

> The last sentence says something about defining device function "foo" via some special option, but there is no further description of this option.

Yes. @kbobrovs mentioned `fsycl-add-targets` option when suggested this use case (https://github.com/intel/llvm/pull/3210#discussion_r575756500). This option allows to supply fat binary with additional device binaries. However this option won't work out of the box now because dynamic linking (and also some other features like spec constants) also require additional information to be generated for each device image and there is no way to add this info through `-fsycl-add-targets` option. That is why I didn't mention the name of this option.

> Does this design also propose some DPC++ extension that allows a new way of defining device functions?

No, it doesn't. The point of this example was that we can actually use dynamic linking not only when shared libraries are involved, but for such use cases when user created device image separately and then supplied the fat binary with it. Such possibility to supply fat binary with some device image existed some time ago and I would consider it as a feature of our implementation rather than language feature.

Reply:
OK, thanks for the explanation.